File: algorithms.py
import math
import sys
from collections.abc import Generator
from typing import Callable

from data import Labyrinth, CellType, Location, Node
from data import SearchState, NodeStack, NodeQueue, NodePrioQueue


def search1(lab: Labyrinth, start: Node, state: SearchState, length_skip: Callable[[],int]) -> Generator[SearchState]:
    state.frontiers.push(start)
    state.frontier_locations = set[Location]()

    while not len(state.frontiers) == 0:
        curr_len = length_skip()
        for _ in range(curr_len):
            # get next node
            current_node = state.frontiers.pop()
            current_location = current_node.location
            state.frontier_locations.discard(current_location)
            state.visited_locations.add(current_location)
            # check current node for goal
            if lab.get_state(current_location).value == CellType.GOAL.value:
                state.result = current_node
                yield state
                return

            # get neighbors and push onto frontier
            neighbors = lab.get_neighbors(current_location)

            for neighbor in neighbors:
                if neighbor not in state.visited_locations and neighbor not in state.frontier_locations:
                    state.frontier_locations.add(neighbor)
                    state.frontiers.push(Node(neighbor, current_node))
        yield state

    yield state
    return

def depth_first_search(lab: Labyrinth, start: Node) -> Generator[SearchState]:
    state = SearchState(NodeStack())
    return search1(lab, start, state, lambda: int(math.sqrt(len(state.frontiers))) + 1)

def breadth_first_search(lab: Labyrinth, start: Node) -> Generator[SearchState]:
    state = SearchState(NodeQueue())
    return search1(lab, start, state, lambda: len(state.frontiers))

def a_star_search(lab: Labyrinth, start: Node) -> Generator[SearchState]:
    state = SearchState(NodePrioQueue(lab.goal()))
    return search1(lab, start, state, lambda: int(math.sqrt(len(state.frontiers))) + 1)


def main():
    rows = 20 if len(sys.argv) < 2 else int(sys.argv[1])
    cols = 40 if len(sys.argv) < 3 else int(sys.argv[2])

    start = Location(0, 0)
    end = Location(rows - 2, cols - 2)
    lab = Labyrinth(rows, cols, .2, start, end)
    breadth_first_search(lab, Node(start, None))

if __name__ == '__main__':
    main()

File: src/algorithms.ts
import {AlgorithmNode, CellType, Labyrinth, SearchState, NodeStack, NodePrioQueue, PositionSet} from "./data"
import {Queue} from "@datastructures-js/queue";

export type SearchResult = Generator<SearchState, undefined, undefined>;

function* base_search(lab: Labyrinth, start: AlgorithmNode, state: SearchState, length_skip: () => number): SearchResult {
    state.frontiers.push(start);
    state.frontier_positions = new PositionSet();

    while(state.frontiers.size() != 0) {
        const current_length = length_skip()

        for (let i = 0; i < current_length; i++) {
            const current_node = state.frontiers.pop();
            const current_position = current_node.position;

            state.frontier_positions.delete(current_position);
            state.visited_positions.add(current_position);

            if(lab.get_state(current_position) == CellType.GOAL) {
                state.result = current_node;
                yield state;
                return;
            }

            const neighbours = lab.get_neighbours(current_position);

            for(const neighbourPosition of neighbours) {
                if(!state.visited_positions.has(neighbourPosition) && !state.frontier_positions.has(neighbourPosition)) {
                    state.frontier_positions.add(neighbourPosition);
                    state.frontiers.push(new AlgorithmNode(neighbourPosition, current_node))
                }
            }
        }
        yield state
    }

    yield state
    return;
}

export function depth_first_search(lab: Labyrinth, start: AlgorithmNode): SearchResult  {
    const state = new SearchState(new NodeStack())

    return base_search(lab, start, state, () => Math.floor(Math.sqrt(state.frontiers.size()))+1)
}

export function breadth_first_search(lab: Labyrinth, start: AlgorithmNode): SearchResult {
    const state = new SearchState(new Queue())

    return base_search(lab, start, state, () => state.frontiers.size());
}

export function a_star_search(lab: Labyrinth, start: AlgorithmNode): SearchResult {
    const state = new SearchState(new NodePrioQueue(lab.goal));
    return base_search(lab, start, state, () => Math.floor(Math.sqrt(state.frontiers.size()))+1);
}

export type AlgorithmFunction = (lab: Labyrinth, start: AlgorithmNode) => SearchResult
export type AlgorithmList = [AlgorithmFunction, string][]

export const ALGORITHM_LIST: AlgorithmList = [
    [breadth_first_search, "Breadth First Search"],
    [depth_first_search, "Depth First Search"],
    [a_star_search, "A* Search"],
]